plugins/git/src/git.rs

```diff
@@ -0,0 +1,267 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::data::*;
+
+use anyhow::Context;
+use anyhow::Result;
+use gix::bstr::ByteSlice;
+use gix::diff::blob::intern::InternedInput;
+use gix::diff::blob::sink::Counter;
+use gix::diff::blob::sources::lines_with_terminator;
+use gix::diff::blob::Algorithm;
+use gix::diff::blob::UnifiedDiffBuilder;
+use gix::object;
+use gix::objs::tree::EntryKind;
+use gix::revision::walk::Sorting;
+use gix::revision::Walk;
+use gix::traverse::commit::simple::CommitTimeOrder;
+use gix::Repository;
+use jiff::Timestamp;
+use std::path::Path;
+use std::sync::OnceLock;
+
+// a lot of the data related to commits can be derived from a Vec<RawCommit>, so cache this and reuse it when possible
+static ALL_RAW_COMMITS: OnceLock<Result<Vec<RawCommit>>> = OnceLock::new();
```

Comment:
Aren't all of these commits tied to a specific repo? The `git` plugin needs to be idempotent. Seems like if `get_all_raw_commits` is called on Repo A, then on Repo B, the execution for Repo B will just re-use the commits from Repo A.

Reply:
Instead you probably want to store a static `(PathBuf, HeadCommit, Vec<RawCommit>)` and use that if possible, but invalidate and replace it when a new `repo` is requested.

Reply:
I can look at it, but I'm pretty sure OnceLock runs once, each time the binary is invoked. Since the `git` plugin is invoked every time `hc check` is run, it will get populated once for each repo. If it wasn't working, then the number of commits reported would be the same on two different repos, as `get_all_raw_commits` is called internally by multiple functions

Reply:
That's under our current usage of the `git` plugin, but someone else might write and run an analysis plugin that doesn't know about this hidden constraint on the `git` plugin. We should design it to be an optimization, but not restrict how the different `git` plugin API endpoints get called.

Reply:
Ok sounds good. I will fix this tomorrow